buildSrc/src/main/kotlin/com/datadog/gradle/plugin/apisurface/KotlinFileVisitor.kt

```diff
@@ -0,0 +1,306 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-2020 Datadog, Inc.
+ */
+
+package com.datadog.gradle.plugin.apisurface
+
+import java.io.File
+import java.lang.UnsupportedOperationException
+import kastree.ast.Node
+import kastree.ast.psi.Parser
+
+class KotlinFileVisitor {
+
+    val description = StringBuilder()
+    private val imports = mutableMapOf<String, String>()
+
+    // region KotlinFileVisitor
+
+    fun visitFile(file: File) {
+        val code = file.readText()
+        val ast = Parser.parseFile(code, throwOnError = false)
+
+        imports.clear()
+        ast.imports.forEach {
+            imports[it.alias ?: it.names.last()] = it.names.joinToString(".")
+        }
+
+        val pkg = ast.pkg?.names?.joinToString(".", postfix = ".").orEmpty()
+        ast.decls.forEach { visitDeclaration(it, pkg, 0) }
+    }
+
+    // endregion
+
+    // region Internal/Visitor
+
+    private fun visitDeclaration(
+        declaration: Node.Decl,
+        pkg: String,
+        level: Int
+    ) {
+        if (declaration is Node.WithModifiers) {
+            if (!declaration.isPublic()) return
+
+            description.append("  ".repeat(level))
+
+            if (declaration.isProtected()) {
+                description.append("protected ")
+            }
+            if (declaration.isOpen()) {
+                description.append("open ")
+            }
+            if (declaration.isAbstract()) {
+                description.append("abstract ")
+            }
+        }
+
+        when (declaration) {
+            is Node.Decl.Structured -> visitStructured(declaration, pkg, level)
+            is Node.Decl.Init -> {}
+            is Node.Decl.Func -> visitFunction(declaration)
+            is Node.Decl.Property -> visitProperty(declaration)
+            is Node.Decl.TypeAlias -> visitTypeAlias(declaration, level)
+            is Node.Decl.Constructor -> visitConstructor(declaration)
+            is Node.Decl.EnumEntry -> visitEnumEntry(declaration)
+        }
+    }
+
+    private fun visitEnumEntry(declaration: Node.Decl.EnumEntry) {
+        description.append("- ")
+        description.append(declaration.name)
+        description.append("\n")
+    }
+
+    private fun visitConstructor(constructor: Node.Decl.Constructor) {
+        description.append("constructor(")
+        visitParameters(constructor.params)
+        description.append(")\n")
+    }
+
+    private fun visitStructured(
```

Comment:
Can you extract some functions and make this function a bit more readable please. It is very hard to follow now given its size.

Reply:
Done